framework/src/base/ComputeJacobianThread.C

```diff
@@ -59,9 +61,39 @@ ComputeJacobianThread::~ComputeJacobianThread()
 void
 ComputeJacobianThread::computeJacobian()
 {
-  if (_kernels.hasActiveBlockObjects(_subdomain, _tid))
+  const MooseObjectWarehouse<KernelBase> * warehouse = NULL;
+  switch (_kernel_type)
   {
-    const std::vector<MooseSharedPointer<KernelBase> > & kernels = _kernels.getActiveBlockObjects(_subdomain, _tid);
+  case Moose::KT_ALL:
+    warehouse = &_nl.getKernelWarehouse();
+    break;
+
+  case Moose::KT_TIME:
+    warehouse = &_nl.getTimeKernelWarehouse();
+    break;
+
+  case Moose::KT_NONTIME:
+    warehouse = &_nl.getNonTimeKernelWarehouse();
+    break;
+
+  case Moose::KT_EIGEN:
+    warehouse = &_nl.getEigenKernelWarehouse();
+    break;
+
+  case Moose::KT_NONEIGEN:
+    warehouse = &_nl.getNonEigenKernelWarehouse();
+    break;
+
+  default:
+    mooseError("Unknown kernel type \n");
+  }
+
+  if (!warehouse)
+    mooseError("Warehouse is NULL \n");
```

Comment:
I do not know if you had this in your previous version, but the only way how `warehouse == NULL` here is that somebody will add a new `KernelType` and implements and "empty" branch in the above switch statement. Note that with new `KernelType` and no branch for it, the code will stop on that `mooseError` in the `default` branch of the switch above...